README.md
# management-api

[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg?style=for-the-badge)](https://github.com/nuvlabox/management-api/graphs/commit-activity)


[![CI](https://img.shields.io/travis/com/nuvlabox/management-api?style=for-the-badge&logo=travis-ci&logoColor=white)](https://travis-ci.com/nuvlabox/management-api)
[![GitHub issues](https://img.shields.io/github/issues/nuvlabox/management-api?style=for-the-badge&logo=github&logoColor=white)](https://GitHub.com/nuvlabox/management-api/issues/)
[![Docker pulls](https://img.shields.io/docker/pulls/nuvlabox/management-api?style=for-the-badge&logo=Docker&logoColor=white)](https://cloud.docker.com/u/nuvlabox/repository/docker/nuvlabox/management-api)
[![Docker image size](https://img.shields.io/microbadger/image-size/nuvlabox/management-api?style=for-the-badge&logo=Docker&logoColor=white)](https://cloud.docker.com/u/nuvlabox/repository/docker/nuvlabox/management-api)

![logo](https://camo.githubusercontent.com/5f893cf6632a9d635c0bdb1c0b51fc97317ce498/68747470733a2f2f6d656469612e73697873712e636f6d2f68756266732f53697853715f47656e6572616c2f6e75766c61626f785f6c6f676f5f7265645f6f6e5f7472616e73706172656e745f3235303070782e706e67)



**This repository contains the source code for the NuvlaBox Management API - this microservice provides the management API for the [NuvlaBox](https://sixsq.com/products-and-services/nuvlabox/overview)**

This microservice is an integral component of the NuvlaBox Engine.

---

**NOTE:** this microservice is part of a loosely coupled architecture, thus when deployed by itself, it might not provide all of its functionalities. Please refer to https://github.com/nuvlabox/deployment for a fully functional deployment

---

## Build the NuvlaBox Management API

This repository is already linked with Travis CI, so with every commit, a new Docker image is released. 

There is a [POM file](pom.xml) which is responsible for handling the multi-architecture and stage-specific builds.

**If you're developing and testing locally in your own machine**, simply run `docker build .` or even deploy the microservice via the local [compose files](docker-compose.yml) to have your changes built into a new Docker image, and saved into your local filesystem.

**If you're developing in a non-master branch**, please push your changes to the respective branch, and wait for Travis CI to finish the automated build. You'll find your Docker image in the [nuvladev](https://hub.docker.com/u/nuvladev) organization in Docker hub, names as _nuvladev/management-api:\<branch\>_.

## Deploy the NuvlaBox Management API

### Prerequisites 

 - *Docker (version 18 or higher)*
 - *Docker Compose (version 1.23.2 or higher)*

### Launching the NuvlaBox Management API

Simply run `docker-compose up --build`

## Contributing

This is an open-source project, so all community contributions are more than welcome. Please read [CONTRIBUTING.md](CONTRIBUTING.md)
 
## Copyright

Copyright &copy; 2019, SixSq Sàrl




CHANGELOG.md
# Changelog
## [1.2.0] - 2021-02-12
### Added
### Changed
 - deprecate HOST_USER in favor of HOME
## [1.1.1] - 2021-01-12
### Added
### Changed
 - ensure existing streaming containers are deleted on duplicated requests
## [1.1.0] - 2021-01-12
### Added 
 - upgraded data-source-mjpg container 
 - additional logging
### Changed
 - reduced number of api workers
## [1.0.0] - 2021-01-08
### Added 
 - auto-restart on docker-proxy failure
### Changed
## [0.4.0] - 2020-12-04
### Added 
  - re-use persisted env vars
### Changed
  - minor fixes
## [0.3.2] - 2020-10-02
### Added 
- ONBUILD SixSq license dump
### Changed
## [0.3.1] - 2020-08-10
### Added
### Changed
- removed file logging
## [0.3.0] - 2020-07-28
### Added 
- new API call for restarting a video stream
### Changed
- refactored functions that are re-used in multiple API calls
## [0.2.0] - 2020-06-23
### Added 
- New API calls and workflow to manage SSH keys on the host
### Changed
## [0.1.0] - 2020-03-13
### Added 
- API endpoints for enabling and disabling MJPG streaming from USB devices
### Changed
## [0.0.2] - 2020-02-18
### Added
### Changed
- re-use compute-api ssl certificates instead of generating our own

## [0.0.1] - 2020-02-17
### Added
- first minimal version of the mgmt api
### Changed


code/app.py
#!/usr/local/bin/python
# -*- coding: utf-8 -*-

"""NuvlaBox Management API

This service provides a management API to the NuvlaBox,
that can be remotely and securely used from Nuvla

Arguments:

"""

import logging
import os
import subprocess
import multiprocessing
import time
import json
import nuvla
from flask import Flask, redirect, request, jsonify, url_for
from management_api.common import utils
from management_api import Manage
from threading import Thread


__copyright__ = "Copyright (C) 2020 SixSq"
__email__ = "support@sixsq.com"

app = Flask(__name__)
app.config["TEMPLATES_AUTO_RELOAD"] = True

logging.basicConfig(format='%(levelname)s - %(funcName)s - %(message)s', level='INFO')
log = logging.getLogger(__name__)


def check_authorized_keys_file():
    """ Makes the initial checks for SSH key management
    Checks if the authorized keys file exists and gives back its content and path

    :returns authorized_keys_file_content, authorized_keys_file_path: content and path of the authorized keys file"""

    authorized_keys_file = "{}/authorized_keys".format(utils.host_ssh_folder)

    if os.path.isfile(authorized_keys_file):
        with open(authorized_keys_file) as akr:
            authorized_keys = akr.read()
    else:
        authorized_keys = ''

    return authorized_keys, authorized_keys_file


def add_ssh_key(pubkey):
    """ Adds a public SSH key to the host's root authorized keys

    :param pubkey: string containing the full public key
    """

    authorized_keys, authorized_keys_file = check_authorized_keys_file()

    with open(authorized_keys_file, 'a+') as ak:
        keys = pubkey.replace('\\n', '\n').splitlines()
        for key in keys:
            if key not in authorized_keys:
                ak.write("\n{}\n".format(key))
                log.info("SSH public key added to host user {}: {}".format(utils.ssh_user, key))
            else:
                log.info("SSH public key {} already added to host. Skipping it".format(key))


def remove_ssh_key(pubkey):
    """ Removes a public SSH key from the host's authorized keys

    :param pubkey: string containing the full public key
    """

    authorized_keys, authorized_keys_file = check_authorized_keys_file()

    # in case the passed value has more than 1 public SSH key
    revoke_keys = pubkey.replace('\\n', '\n').splitlines()
    final_keys = authorized_keys.splitlines()
    for key in revoke_keys:
        if key in final_keys:
            final_keys.remove(key)

    if final_keys != authorized_keys.splitlines():
        with open(authorized_keys_file, 'w') as ak:
            ak.write("\n".join(final_keys))
        log.info("SSH public key removed from host user {}: {}".format(utils.ssh_user, pubkey))
    else:
        log.info("The provided SSH public key {} is not in the host's authorized keys. Nothing to do".format(pubkey))


def default_ssh_key():
    """ Looks for the env var NUVLABOX_SSH_PUB_KEY, and add the respective
     SSH public key to the host
    """

    if utils.provided_pubkey:
        log.info("Environment variable NUVLABOX_SSH_PUB_KEY found. Adding key to host user {}".format(utils.ssh_user))
        add_ssh_key(utils.provided_pubkey)


def wait_for_certificates():
    """ If there are already TLS credentials for the compute-api, then re-use them """

    log.info("Re-using compute-api SSL certificates for NuvlaBox Management API")
    log.info("Waiting for compute-api to generate SSL certificates...")

    while not os.path.exists("{}/{}".format(utils.nuvlabox_api_certs_folder, utils.server_cert_file)) or \
            not os.path.exists("{}/{}".format(utils.nuvlabox_api_certs_folder, utils.server_key_file)) or \
            not os.path.exists("{}/{}".format(utils.nuvlabox_api_certs_folder, utils.ca_file)):

        time.sleep(3)


def request_stop_mjpg_streamer_container(name, nuvla_resource_id):
    log.info("Stopping container {}".format(name))
    Manage.stop_container_data_source_mjpg(name)
    try:
        log.info("Updating {} in Nuvla".format(nuvla_resource_id))
        Manage.update_peripheral_resource(nuvla_resource_id, data_gateway_enabled=False)
    except nuvla.api.api.NuvlaError as e:
        if e.response.status_code == 404:
            log.warning("Peripheral {} has already been delete in Nuvla. Nothing to do".format(nuvla_resource_id))
            # this action was triggered by the deletion of the peripheral,
            # so it's normal that it does not exist anymore
            pass
        else:
            log.exception("Could not update {} in Nuvla. Trying a second time...".format(nuvla_resource_id))
            # try again. If still fails, then raise the exception
            Manage.update_peripheral_resource(nuvla_resource_id, data_gateway_enabled=False)


def request_start_mjpg_streamer_container(name, nuvla_resource_id, device, resolution, fps):
    log.info("Launching MJPG streamer container {} for {}".format(name, device))
    local_data_gateway_endpoint, container = Manage.start_container_data_source_mjpg(name,
                                                                                     device,
                                                                                     resolution,
                                                                                     fps)
    if container.status.lower() == 'created':
        log.info("MJPG streamer {} successfully created. Updating {} in Nuvla".format(name, nuvla_resource_id))

        Manage.update_peripheral_resource(nuvla_resource_id, local_data_gateway_endpoint=local_data_gateway_endpoint)
        return True, container.logs().decode('utf-8')
    else:
        log.error("MJPG streamer {} could not be started: {}".format(name, container.status))

        return False, container.logs().decode('utf-8')


@app.errorhandler(404)
def page_not_found(e):
    # return a JSON error code, explicit
    return jsonify(dict(utils.return_404, message=str(e))), utils.return_404['status']


@app.errorhandler(405)
def method_not_allowed(e):
    # return a JSON error code, explicit
    return jsonify(dict(utils.return_405, message=str(e))), utils.return_405['status']


@app.route("/")
def root():
    # redirect to self-discovery api endpoint
    return redirect("/api", code=302)


@app.route("/api")
def self_discovery():
    # return a list of all api endpoints
    links = []
    for rule in app.url_map.iter_rules():
        if rule.endpoint != 'static':
            url = url_for(rule.endpoint, **(rule.defaults or {}))
            links.append(url)
    return jsonify({"nuvlabox-api-endpoints": links}), 200


@app.route("/api/reboot", methods=['POST'])
def reboot():
    # reboot the host

    thread = Thread(target=Manage.reboot)
    thread.start()

    return jsonify(dict(utils.return_200, message="rebooting NuvlaBox machine...")), utils.return_200['status']


@app.route("/api/add-ssh-key", methods=['POST'])
def accept_new_ssh_key():
    # adds an SSH key into the host's authorized keys
    # the payload is the public key is, raw
    payload = request.data.decode('UTF-8')

    log.info("Received request to add public SSH key to host: {}".format(payload))

    if not payload or not isinstance(payload, str):
        return jsonify(dict(utils.return_400, message="Payload should match a valid public SSH key. Recevied: %s" %
                                                      payload)), \
               utils.return_400['status']

    try:
        add_ssh_key(payload)
        return jsonify(dict(utils.return_200, message="Added public SSH key to host: {}".format(payload))), \
               utils.return_200['status']
    except Exception as e:
        log.exception("Cannot add public SSH key to host: {}".format(e))
        if e.status_code:
            return jsonify(dict(utils.return_500, message=str(e), status=e.status_code)), e.status_code
        else:
            return jsonify(dict(utils.return_500, message=str(e))), utils.return_500['status']


@app.route("/api/revoke-ssh-key", methods=['POST'])
def revoke_ssh_key():
    # removes the SSH public key passed in the payload,
    # from the host's authorized keys
    payload = request.data.decode('UTF-8')

    log.info("Received request to revoke public SSH key from host: {}".format(payload))

    if not payload or not isinstance(payload, str):
        return jsonify(dict(utils.return_400, message="Payload should match a valid public SSH key. Recevied: %s" %
                                                      payload)), \
               utils.return_400['status']

    try:
        remove_ssh_key(payload)
        return jsonify(dict(utils.return_200, message="Removed SSH key from host: {}".format(payload))), \
               utils.return_200['status']
    except Exception as e:
        log.exception("Cannot revoke public SSH key from host: {}".format(e))
        if e.status_code:
            return jsonify(dict(utils.return_500, message=str(e), status=e.status_code)), e.status_code
        else:
            return jsonify(dict(utils.return_500, message=str(e))), utils.return_500['status']


@app.route("/api/data-source-mjpg/enable", methods=['POST'])
def enable_data_source_mjpg():
    # enable data gateway for mjpg
    #
    # payload looks like:
    # { "id": str, "resolution": str, "fps": int, "video-device": str }
    payload = json.loads(request.data)
    mandatory_keys = {"id", "video-device"}

    if not mandatory_keys <= set(payload.keys()):
        return jsonify(dict(utils.return_400, message="Missing mandatory attributes - %s" % mandatory_keys)), \
               utils.return_400['status']

    name = payload['id'].split("/")[-1]

    try:
        resolution = payload['resolution']
    except KeyError:
        resolution = "1280x720"

    try:
        fps = int(payload['fps'])
    except KeyError:
        fps = 15

    log.info("Received /api/data-source-mjpg/enable request with payload: {}".format(payload))

    try:
        success, container_logs = request_start_mjpg_streamer_container(name, payload['id'],
                                                                        payload['video-device'],
                                                                        resolution,
                                                                        fps)

        if success:
            return jsonify(dict(utils.return_200, message="MJPG streamer started!")), utils.return_200['status']
        else:
            return jsonify(dict(utils.return_400, message=container_logs)), utils.return_400['status']
    except Exception as e:
        log.exception("Cannot enable stream: {}".format(e))
        if e.status_code:
            return jsonify(dict(utils.return_500, message=str(e), status=e.status_code)), e.status_code
        else:
            return jsonify(dict(utils.return_500, message=str(e))), utils.return_500['status']


@app.route("/api/data-source-mjpg/disable", methods=['POST'])
def disable_data_source_mjpg():
    # disable data gateway for mjpg
    #
    # payload looks like:
    # { "id": str }
    payload = json.loads(request.data)
    mandatory_keys = {"id"}

    if not mandatory_keys <= set(payload.keys()):
        return jsonify(dict(utils.return_400, message="Missing mandatory attributes - %s" % mandatory_keys)), \
               utils.return_400['status']

    name = payload['id'].split("/")[-1]

    log.info("Received /api/data-source-mjpg/disable request with payload: {}".format(payload))

    try:
        request_stop_mjpg_streamer_container(name, payload['id'])

        return jsonify(dict(utils.return_200, message="MJPG streamer stopped for %s" % payload['id'])), \
               utils.return_200['status']
    except Exception as e:
        log.exception("Cannot disable stream: {}".format(e))
        if e.status_code:
            return jsonify(dict(utils.return_500, message=str(e), status=e.status_code)), e.status_code
        else:
            return jsonify(dict(utils.return_500, message=str(e))), utils.return_500['status']


@app.route("/api/data-source-mjpg/restart", methods=['POST'])
def restart_data_source_mjpg():
    # restart data gateway for mjpg
    #
    # payload looks like:
    # { "id": str, "video-device": str }
    payload = json.loads(request.data)
    mandatory_keys = {"id", "video-device"}

    if not mandatory_keys <= set(payload.keys()):
        return jsonify(dict(utils.return_400, message="Missing mandatory attributes - %s" % mandatory_keys)), \
               utils.return_400['status']

    name = payload['id'].split("/")[-1]

    get_env = Manage.find_container_env_vars(name, keys=["RESOLUTION", "FPS"])
    resolution = get_env.get("RESOLUTION", "1280x720")
    fps = get_env.get("FPS", 15)

    log.info("Received /api/data-source-mjpg/restart request with payload: {}".format(payload))

    try:
        request_stop_mjpg_streamer_container(name, payload['id'])
        success, container_logs = request_start_mjpg_streamer_container(name, payload['id'],
                                                                        payload['video-device'],
                                                                        resolution,
                                                                        fps)

        if success:
            return jsonify(dict(utils.return_200, message="MJPG streamer restarted!")), utils.return_200['status']
        else:
            return jsonify(dict(utils.return_400, message=container_logs)), utils.return_400['status']
    except Exception as e:
        log.exception("Cannot restart stream (old streamer might be in a faulty state): {}".format(e))
        if e.status_code:
            return jsonify(dict(utils.return_500, message=str(e), status=e.status_code)), e.status_code
        else:
            return jsonify(dict(utils.return_500, message=str(e))), utils.return_500['status']


if __name__ == "__main__":
    """ Main """

    # Check if there is an SSH key to be added to the host
    try:
        default_ssh_key()
    except:
        # it is not critical if we can't add it, for any reason
        log.exception("Could not add NUVLABOX_SSH_PUB_KEY to the host root. Moving on and discarding the provided key")

    # Let's re-use the certificates already generated for the compute-api
    wait_for_certificates()

    log.info("Starting NuvlaBox Management API!")
    try:
        subprocess.check_output(["gunicorn", "--bind=0.0.0.0:5001", "--threads=2",
                                 "--worker-class=gthread", "--workers=1", "--reload",
                                 "--keyfile", "{}/{}".format(utils.nuvlabox_api_certs_folder, utils.server_key_file),
                                 "--certfile", "{}/{}".format(utils.nuvlabox_api_certs_folder, utils.server_cert_file),
                                 "--ca-certs", "{}/{}".format(utils.nuvlabox_api_certs_folder, utils.ca_file),
                                 "--cert-reqs", "2", "--no-sendfile",
                                 "wsgi:app"])
    except FileNotFoundError:
        log.exception("Gunicorn not available!")
        raise
    except (OSError, subprocess.CalledProcessError):
        log.exception("Failed start NuvlaBox Management API!")
        log.exception("Failed start NuvlaBox Management API!")
        raise



code/management_api/Manage.py
#!/usr/local/bin/python3.7
# -*- coding: utf-8 -*-

""" All the management functions that can be called from the API """

import os
import time
import docker
import json
from management_api.common import utils
from nuvla.api import Api

# TODO: is there a way to avoid fixing a tag? If a new tag comes out, we need to update this microservice
data_gateway_images = {
    "data_source_mjpg": "nuvlabox/data-source-mjpg:0.0.2"
}


def reboot():
    # reboot the host
    # NOTE: there's no return from this function
    time.sleep(5)
    os.system("echo b > /sysrq")


def enable_ssh():
    # TODO
    time.sleep(5)


def start_container_data_source_mjpg(name, video_device, resolution, fps):
    """ starts mjpg streamer

    :param name: unique name of the container and termination of the pathprefix in traefik
    :param video_device: path to video device, i.e. /dev/video0
    :param resolution: video feed resolution (str like WxH)
    :param fps: number of frames per second

    :returns local_data_gateway_endpoint and container obj
    """
    client = docker.from_env()

    try:
        existing_container = client.containers.get(name)
        if existing_container:
            # we force kill any previous container, if there's a new request for the same streamer
            existing_container.remove(force=True)
    except docker.errors.NotFound:
        # this is good, the container shouldn't exist
        pass
    except:
        # better not move forward, as it can cause zombie containers
        raise

    cmd = '--input-type input_uvc.so --device-path {} --resolution {} --fps {}'.format(video_device,
                                                                                       resolution,
                                                                                       fps)

    path_prefix = "/video/{}".format(name)

    devices = ["{}:{}".format(video_device, video_device)]

    traefik_router_name = "traefik.http.routers.mjpg-streamer-{}-router".format(name)

    labels = {"nuvlabox.component": "True",
              "nuvlabox.data-source-container": "True",
              "traefik.enable": "true",
              "{}.rule".format(traefik_router_name): "Host(`data-gateway`) && PathPrefix(`{}`)".format(path_prefix),
              "{}.entrypoints".format(traefik_router_name): "web",
              "{}.middlewares".format(traefik_router_name): "{}-mid".format(name),
              "traefik.http.services.mjpg-streamer-{}-service.loadbalancer.server.port".format(name): "8082",
              "traefik.http.middlewares.{}-mid.replacepath.path".format(name): "/"
              }

    streaming_url = 'http://data-gateway' + path_prefix + '?action=stream'
    return streaming_url, client.containers.run(data_gateway_images['data_source_mjpg'],
                                                command=cmd,
                                                detach=True,
                                                name=name,
                                                hostname=name,
                                                devices=devices,
                                                labels=labels,
                                                network="nuvlabox-shared-network",
                                                restart_policy={"Name": "always"},
                                                environment=[f"RESOLUTION={resolution}",
                                                             f"FPS={fps}",
                                                             "CRON_DATAGATEWAY_HEALTHCHECK=1"]
                                                )


def stop_container_data_source_mjpg(name):
    """ stops mjpg streamer container by name

    :param name: unique name of the container

    :returns
    """
    client = docker.from_env()

    try:
        client.containers.get(name).remove(force=True)
    except docker.errors.NotFound:
        pass


def nuvla_api():
    """ Initialize API instance """
    if os.path.exists(utils.nuvla_configuration):
        nuvla_endpoint_raw = nuvla_endpoint_insecure_raw = None
        with open(utils.nuvla_configuration) as nuvla_conf:
            for line in nuvla_conf.read().split():
                try:
                    if line and 'NUVLA_ENDPOINT=' in line:
                        nuvla_endpoint_raw = line.split('=')[-1]
                    if line and 'NUVLA_ENDPOINT_INSECURE=' in line:
                        nuvla_endpoint_insecure_raw = bool(line.split('=')[-1])
                except IndexError:
                    pass

        if nuvla_endpoint_raw and nuvla_endpoint_insecure_raw:
            api = Api(endpoint='https://{}'.format(nuvla_endpoint_raw),
                      insecure=nuvla_endpoint_insecure_raw, reauthenticate=True)
        else:
            raise Exception(f'Misconfigured Nuvla parameters in {utils.nuvla_configuration}. Cannot perform operation')
    else:
        raise Exception("NuvlaBox is not yet ready to be operated. Missing Nuvla configuration parameters")

    try:
        with open(utils.activation_flag) as a:
            user_info = json.loads(a.read())
    except FileNotFoundError:
        raise Exception("Cannot authenticate back with Nuvla")

    api.login_apikey(user_info['api-key'], user_info['secret-key'])

    return api


def find_container_env_vars(container_name, keys=None):
    """ Inspects a container and looks up its environment variables
    If the arg keys is passed, it then looks up the values for those keys. Otherwise, it returns the full env

    :param container_name: container to be inspected
    :param keys: list of env vars to look up. If not passed, it will look up the whole environment
    :returns {key1: value1, key2: value2} - a map of the provided keys and respective values in the container env
    """

    client = docker.from_env()

    insp = client.api.inspect_container(container_name)

    try:
        env = insp['Config']['Env']
    except KeyError:
        env = []

    env_map = {}
    if keys:
        for k in keys:
            try:
                var = list(filter(lambda x: f"{k}=" in x, env))[0]
                env_map[var.split("=", 1)[0]] = var.split("=", 1)[1]
            except IndexError:
                continue
    else:
        for env_var in env:
            env_map[env_var.split("=", 1)[0]] = env_var.split("=", 1)[1]

    return env_map


def update_peripheral_resource(id, local_data_gateway_endpoint=None, data_gateway_enabled=True, raw_sample=None):
    """ sends a PUT request to Nuvla to update the peripheral resource whenever a data gateway action takes place

    :param id: UUID of the peripheral resource in nuvla
    :param local_data_gateway_endpoint: data gateway url for accessing the routed data
    :param data_gateway_enabled: whether data dateway is enabled or not
    :param raw_sample: raw data sample

    :returns """

    api = nuvla_api()
    kwargs = {'select': []}

    payload = {
        "data-gateway-enabled": data_gateway_enabled
    }

    if local_data_gateway_endpoint:
        payload['local-data-gateway-endpoint'] = local_data_gateway_endpoint
    else:
        kwargs['select'].append("local-data-gateway-endpoint")

    if raw_sample:
        payload['raw-data-sample'] = raw_sample
    else:
        kwargs['select'].append("raw-data-sample")

    api._cimi_put(id, json=payload, params=kwargs)


code/management_api/common/utils.py
#!/usr/local/bin/python3.7
# -*- coding: utf-8 -*-

""" Common set of managament methods to be used by
 the different management api classes """

import os

# nuvla_endpoint_raw = os.environ["NUVLA_ENDPOINT"] if "NUVLA_ENDPOINT" in os.environ else "nuvla.io"
# while nuvla_endpoint_raw[-1] == "/":
#     nuvla_endpoint_raw = nuvla_endpoint_raw[:-1]
#
# nuvla_endpoint_insecure_raw = os.environ["NUVLA_ENDPOINT_INSECURE"] if "NUVLA_ENDPOINT_INSECURE" in os.environ else False
# if isinstance(nuvla_endpoint_insecure_raw, str):
#     if nuvla_endpoint_insecure_raw.lower() == "false":
#         nuvla_endpoint_insecure_raw = False
#     else:
#         nuvla_endpoint_insecure_raw = True
# else:
#     nuvla_endpoint_insecure_raw = bool(nuvla_endpoint_insecure_raw)
#
# nuvla_endpoint_insecure = nuvla_endpoint_insecure_raw
#
# nuvla_endpoint = nuvla_endpoint_raw.replace("https://", "")

data_volume = "/srv/nuvlabox/shared"
log_filename = "management-api.log"

activation_flag = "{}/.activated".format(data_volume)
nuvla_configuration = f'{data_volume}/.nuvla-configuration'

server_cert_file = "server-cert.pem"
server_key_file = "server-key.pem"
client_cert_file = "cert.pem"
client_key_file = "key.pem"
ca_file = "ca.pem"

nuvlabox_api_certs_folder = data_volume

provided_pubkey = os.getenv("NUVLABOX_SSH_PUB_KEY")
host_ssh_folder = "/hostfs/.ssh"

host_home_user = os.getenv("HOST_USER", os.getenv('HOME'))
ssh_user = host_home_user if host_home_user else "root"

return_404 = {"status": 404,
              "message": "undefined"}

return_405 = {"status": 405,
              "message": "undefined"}

return_200 = {"status": 200,
              "message": "undefined"}

return_201 = {"status": 201,
              "message": "undefined"}

return_400 = {"status": 400,
              "message": "undefined"}

return_500 = {"status": 500,
              "message": "undefined"}

return_generic = {"status": "placeholder",
                  "message": "undefined"}

